custom_components/mammotion/lawn_mower.py

```diff
@@ -232,3 +236,42 @@ async def async_pause(self) -> None:
             self.coordinator.async_set_updated_data(
                 self.coordinator.manager.mower(self.coordinator.device_name)
             )
+
+    async def async_cancel(self) -> None:
+        """Cancel Job"""
+        
+        mode = self.rpt_dev_status.sys_status
+        if mode is None:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN, translation_key="device_not_ready"
+            )
+            
+        if mode in (
+            WorkMode.MODE_PAUSE,
+            WorkMode.MODE_WORKING,
+            WorkMode.MODE_RETURNING,
+        ):
+            try:
+                if mode != WorkMode.MODE_PAUSE:
+                    if mode == WorkMode.MODE_WORKING:
+                        trans_key = "pause_failed"
+                        await self.coordinator.async_send_command("pause_execute_task")
+                    if mode == WorkMode.MODE_RETURNING:
+                        trans_key = "dock_failed"
+                        await self.coordinator.async_send_command("cancel_return_to_dock")
+                    await self.coordinator.async_request_iot_sync()
+                    mode = self.rpt_dev_status.sys_status
+                    
+                if mode == WorkMode.MODE_PAUSE:
+                    trans_key = "pause_failed"
+                    await self.coordinator.async_send_command("cancel_job")
+
+            except COMMAND_EXCEPTIONS as exc:
+                raise HomeAssistantError(
+                    translation_domain=DOMAIN, translation_key=trans_key
+                ) from exc
+            finally:
+                await self.coordinator.async_request_iot_sync()
+                self.coordinator.async_set_updated_data(
```

Comment:
You shouldn't need this

Reply:
all messages returned from MQTT trigger an update, see `_async_update_notification` in coordinator

Reply:
Otherwise looks good for now, this logic needs to move out to the library but will do at the moment.